### ISSUE TITLE
Operator precedence tree compiling extremely slowly

### ISSUE THREAD

**Issue by npip99** [pegjs/pegjs#656]
### Issue type

- **Bug Report:** Runs very slow for particular PEGs

### Description

This particular pegjs file is running very slowly: https://pastebin.com/3N2VwDYz

It appears to directly be caused by the operator precedence tree. I commented out two of the layers to help it run in a reasonable time, and commenting various different numbers of layers allows me to freely adjust how slow pegjs is. A similar operator precedence tree was implemented perfectly fine in the example javascript.pegjs, so it appears that peg is having a problem with the particular way in which I wrote it. Interestingly, the resulting parser is only 4000 lines of javascript and when scrolling through it it looks very straight forward and doesn't appear to have anything that clearly would've caused the choke-up. The lines map up very directly to exactly the pegjs file.

Profiling for the slow .pegjs file: https://pastebin.com/fMY7q10r
Profiling for the fast javascript.pegjs file: https://pastebin.com/V1YMEiGw

I've tracked that using the example javascript.pegjs file, `function visit(node)` on line 11 of `lib/compiler/visitor.js` runs ~220,000 times. However usinig the .pegjs file provided, it runs ~158,000,000 times. Something must be running in O(exponential) time internally.

EDIT: In-fact, yes, we do have O(exponential) internally. Specifically, it grows exponentially with respect to the number of branches. That must be why `javascript.pegjs` runs in O(n) time, because it has no layered branches. By jumping to `precedence_8 = precedence_5` instead of `precedence_8 = precedence_6`, `visitor.js` only runs ~30,000,000 times, thus reducing the number of operations by 1/5th. And in particular, precedence_6 had 5 branches. So clearly each layer of branches is multiplying the run-time by the number of branches in that layer. So it's running in O(product(number of branches)) over the number of layers. I am passing in the cached flag, so something in caching isn't working correctly as subtrees are supposed to be memoized for this type of operation. Now I know how to change my .pegjs file in order to avoid this issue, ie just trim down the branching, but I'll still leave this issue here in-case anyone involved in pegjs ever wants to memoize this.

### Steps to Reproduce

Run the given .pegjs file.

**Expected behavior:**

Compiles within 1s

**Actual behavior:**

Compiles over the course of 15-20 seconds. Noting that the time complexity is O(product(number of branches)) in the README may be helpful for the time being.

### Software

- **PEG.js:** 0.10.0
- **Node.js:** v8.10.0
- **NPM or Yarn:** 3.5.2
- **OS:** Linux (WSL)

**Comment by Mingun** [comment]
Your precedence rules implemented in a very unoptimal way for PEG parsers. Use repetitions instead of choices. See also https://github.com/pegjs/pegjs/issues/538#issuecomment-343547210

**Comment by npip99** [comment]
@Mingun Yep! I know now that I can reorganize my statements to avoid this dilemma and I have since done this, which would then fix the problem. Thanks for the tips. As an aside for the devs, depending on the dev status though, someone might want to fix it. Exponential time is clearly not the theoretically optimal implementation, and scrolling through the emitted .js parser file allows one to see how simply the original pegjs could be translated to a js parser by hand that runs in O(n) time.
